include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -439,44 +440,124 @@ auto applyDeflation(const SizeType i_begin, const SizeType i_end, RhoSender&& rh
+                                 auto& ws_vecs) {
+        const matrix::Distribution distr(LocalElementSize(n, n), TileElementSize(nb, nb));
+
+        const std::size_t batch_size = util::ceilDiv(to_sizet(k), nthreads);
+        const std::size_t begin = thread_idx * batch_size;
+        const std::size_t end = std::min(thread_idx * batch_size + batch_size, to_sizet(k));
+
+        // STEP 0: Initialize workspaces (single-thread)
+        if (thread_idx == 0)
+          ws_vecs.resize(nthreads);
+
+        barrier_ptr->arrive_and_wait();
+
+        // STEP 1: LAED4 (multi-thread)
+        const T* d_ptr = d_tiles_futs[0].get().ptr();
+        const T* z_ptr = z_tiles[0].ptr();
+        T* eval_ptr = eval_tiles[0].ptr();
+
+        for (std::size_t i = begin; i < end; ++i) {
+          T& eigenval = eval_ptr[i];
+
+          const SizeType i_tile = distr.globalTileLinearIndex(GlobalElementIndex(0, to_SizeType(i)));
+          const SizeType i_col = distr.tileElementFromGlobalElement<Coord::Col>(to_SizeType(i));
+          T* delta = evec_tiles[to_sizet(i_tile)].ptr(TileElementIndex(0, i_col));
+
+          lapack::laed4(to_int(k), to_int(i), d_ptr, z_ptr, delta, rho, &eigenval);
+        }
 
-    std::vector<SizeType> loop_arr(to_sizet(k));
-    std::iota(std::begin(loop_arr), std::end(loop_arr), 0);
-    pika::for_each(pika::execution::par, std::begin(loop_arr), std::end(loop_arr), [&](const SizeType i) {
-      T& eigenval = eval_ptr[to_sizet(i)];
+        barrier_ptr->arrive_and_wait();
 
-      const SizeType i_tile = distr.globalTileLinearIndex(GlobalElementIndex(0, i));
-      const SizeType i_col = distr.tileElementFromGlobalElement<Coord::Col>(i);
-      T* delta = evec_tiles[to_sizet(i_tile)].ptr(TileElementIndex(0, i_col));
+        // STEP 2a Compute weights
+        auto& q = evec_tiles;
+        auto& w = ws_vecs[thread_idx];
 
-      lapack::laed4(static_cast<int>(k), static_cast<int>(i), d_ptr, z_ptr, delta, rho, &eigenval);
-    });
-  };
+        // - copy diagonal from q -> w (or just initialize with 1)
+        w.resize(k, 1);
+        if (thread_idx == 0) {
+          for (auto i = 0; i < k; ++i) {
+            const GlobalElementIndex kk(i, i);
+            const auto diag_tile = distr.globalTileLinearIndex(kk);
+            const auto diag_element = distr.tileElementIndex(kk);
 
-  TileCollector tc{i_begin, i_end};
+            w[i] = q[to_sizet(diag_tile)](diag_element);
+          }
+        }
 
-  auto sender =
-      ex::when_all(std::forward<KSender>(k), std::forward<RhoSender>(rho),
-                   ex::when_all_vector(tc.read(d)), ex::when_all_vector(tc.read(z)),
-                   ex::when_all_vector(tc.readwrite(evals)), ex::when_all_vector(tc.readwrite(evecs)));
+        // - compute productorial (thread-local)
+        for (auto j = to_SizeType(begin); j < to_SizeType(end); ++j) {
+          for (auto i = 0; i < k; ++i) {
+            if (i == j)
+              continue;
```

Comment:
`if` statements in inner loops should be avoided (performance problem).
It might have a negligible effect on the full algorithm, but maybe the two loops option should be considered.

If you want to keep this approach please consider to embed ll 500-508 in the `i==j` part (improved code readability)